File: examples/basic-demos/src/hooks.ts
/**
 * Hook implementations for Demo TypeScript CLI
 * 
 * This file contains the business logic for your CLI commands.
 * Implement the hook functions below to handle your CLI commands.
 * 
 * IMPORTANT: Hook names must use snake_case with 'on_' prefix
 * Example:
 * - Command 'hello' -> Hook function 'on_hello'
 * - Command 'hello-world' -> Hook function 'on_hello_world'
 */

// Import any modules you need here
import * as fs from 'fs';
import * as path from 'path';
/**
 * Greet someone with style * @param name Name to greet * @param message Custom greeting message * @param options Command options * @param options.style Greeting style * @param options.count Repeat greeting N times * @param options.uppercase Convert to uppercase * @param options.language Language code * @returns Promise<number> - Exit code (0 for success)
 */
export async function on_greet(    name: string,    message: string,    options: {        style?: string;        count?: string;        uppercase?: string;        language?: string;    }): Promise<number> {    // Implement greeting business logic with TypeScript types
    const nameValue: string = name || 'World';
    const messageValue: string = message || 'Hello';
    const style: string = options?.style || 'casual';
    const count: number = parseInt(String(options?.count)) || 1;
    const uppercase: boolean = options?.uppercase || false;
    const language: string = options?.language || 'en';
    
    // Language-specific greetings with types
    const greetings: { [key: string]: { [style: string]: string } } = {
        'en': { casual: 'Hi', formal: 'Hello', friendly: 'Hey' },
        'es': { casual: 'Hola', formal: 'Buenos días', friendly: 'Ey' },
        'fr': { casual: 'Salut', formal: 'Bonjour', friendly: 'Coucou' },
        'de': { casual: 'Hallo', formal: 'Guten Tag', friendly: 'Hey' }
    };
    
    const greetingWord: string = greetings[language]?.[style] || greetings['en'][style] || messageValue;
    let output: string = `${greetingWord}, ${nameValue}!`;
    
    if (uppercase) {
        output = output.toUpperCase();
    }
    
    // Repeat the greeting
    for (let i = 0; i < count; i++) {
        console.log(output);
    }    
    // Return 0 for success, non-zero for error
    return 0;
}
/**
 * Display system and environment information * @param options Command options * @param options.format Output format * @param options.verbose Show detailed information * @param options.sections Comma-separated sections to show * @returns Promise<number> - Exit code (0 for success)
 */
export async function on_info(    options: {        format?: string;        verbose?: string;        sections?: string;    }): Promise<number> {    // Implement system info display logic with TypeScript types
    const format: string = options?.format || 'text';
    const verbose: boolean = options?.verbose || false;
    const sections: string[] = (options?.sections || 'all').split(',').map((s: string) => s.trim());
    
    interface SystemInfo {
        system: {
            platform: string;
            arch: string;
            version: string;
            uptime: number;
        };
        environment: {
            nodeVersion: string;
            cwd: string;
            user: string;
        };
        memory: {
            used: number;
            heap: number;
            external: number;
        };
    }
    
    const info: SystemInfo = {
        system: {
            platform: process.platform,
            arch: process.arch,
            version: process.version,
            uptime: Math.floor(process.uptime())
        },
        environment: {
            nodeVersion: process.version,
            cwd: process.cwd(),
            user: process.env.USER || process.env.USERNAME || 'unknown'
        },
        memory: {
            used: Math.round(process.memoryUsage().rss / 1024 / 1024),
            heap: Math.round(process.memoryUsage().heapUsed / 1024 / 1024),
            external: Math.round(process.memoryUsage().external / 1024 / 1024)
        }
    };
    
    const showSection = (section: string): boolean => sections.includes('all') || sections.includes(section);
    
    if (format === 'json') {
        const output: Partial<SystemInfo> = {};
        if (showSection('system')) output.system = info.system;
        if (showSection('environment')) output.environment = info.environment;
        if (showSection('memory')) output.memory = info.memory;
        console.log(JSON.stringify(output, null, verbose ? 2 : 0));
    } else {
        console.log('System Information:');
        
        if (showSection('system')) {
            console.log('\n📱 System:');
            console.log(`  Platform: ${info.system.platform}`);
            console.log(`  Architecture: ${info.system.arch}`);
            console.log(`  Node.js: ${info.system.version}`);
            if (verbose) console.log(`  Uptime: ${info.system.uptime}s`);
        }
        
        if (showSection('environment')) {
            console.log('\n🌍 Environment:');
            console.log(`  Working Directory: ${info.environment.cwd}`);
            console.log(`  User: ${info.environment.user}`);
            if (verbose) console.log(`  Node Version: ${info.environment.nodeVersion}`);
        }
        
        if (showSection('memory')) {
            console.log('\n💾 Memory Usage:');
            console.log(`  RSS: ${info.memory.used}MB`);
            console.log(`  Heap Used: ${info.memory.heap}MB`);
            if (verbose) console.log(`  External: ${info.memory.external}MB`);
        }
    }    
    // Return 0 for success, non-zero for error
    return 0;
}

File: examples/basic-demos/src/hooks.rs
//! Hook implementations for Demo Rust Complex CLI
//! 
//! This file contains the business logic for your CLI commands.
//! Implement the hook functions below to handle your CLI commands.
//! 
//! Each command in your CLI corresponds to a hook function.

use clap::ArgMatches;
use anyhow::{Result, anyhow};
use serde_json::json;
use std::collections::HashMap;

use crate::logger::{info, warn, error, debug, update_context, LogContext};

/// Hook function for 'process' command
/// Process data with progress bars and table output
pub fn on_process(matches: &ArgMatches) -> Result<()> {
    // Set up logging context for this command
    let mut context: LogContext = HashMap::new();
    context.insert("command".to_string(), json!("process"));
    update_context(context);

    info("hooks", "Executing process command", None);
    
    // Add your business logic here
    
    // Extract arguments
    if let Some(input) = matches.get_one::<String>("input") {
        println!("input: {}", input);
    }
    
    // Extract options
    if let Some(format) = matches.get_one::<String>("format") {
        println!("format: {}", format);
    }
    let progress = matches.get_flag("progress");
    println!("progress: {}", progress);
    let verbose = matches.get_flag("verbose");
    println!("verbose: {}", verbose);
    
    // Return Ok(()) for success, Err(...) for error
    Ok(())
}

/// Hook function for 'config' command
/// Manage configuration settings
pub fn on_config(matches: &ArgMatches) -> Result<()> {
    // Set up logging context for this command
    let mut context: LogContext = HashMap::new();
    context.insert("command".to_string(), json!("config"));
    update_context(context);

    info("hooks", "Executing config command", None);
    
    // Add your business logic here
    
    // Extract options
    if let Some(get) = matches.get_one::<String>("get") {
        println!("get: {}", get);
    }
    if let Some(set) = matches.get_one::<String>("set") {
        println!("set: {}", set);
    }
    
    // Return Ok(()) for success, Err(...) for error
    Ok(())
}


/// Simple hook manager for compatibility with other templates
pub struct HookManager;

impl HookManager {
    /// Create a new hook manager
    pub fn new() -> Self {
        Self
    }
    
    /// Execute a hook by name
    pub fn execute_hook(&self, name: &str, matches: &ArgMatches) -> Result<()> {
        match name {

            "on_process" => on_process(matches),
            "on_config" => on_config(matches),
            _ => Err(anyhow::anyhow!("Unknown hook: {}", name)),
        }
    }
}

impl Default for HookManager {
    fn default() -> Self {
        Self::new()
    }
}

// Add any utility functions or structures here